File: sdks/python/reshadx/utils/http.py
"""ReshADX Python SDK - HTTP Client"""

import time
from typing import Any, Dict, Optional
from urllib.parse import urljoin

import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

from .errors import ReshADXError, ServerError


class HttpClient:
    """HTTP client with retry logic"""

    def __init__(
        self,
        api_key: str,
        environment: str = "production",
        base_url: Optional[str] = None,
        timeout: int = 30,
        max_retries: int = 3,
    ):
        self.api_key = api_key
        self.access_token: Optional[str] = None
        self.timeout = timeout

        # Determine base URL
        if base_url:
            self.base_url = base_url
        elif environment == "production":
            self.base_url = "https://api.reshadx.com/v1"
        else:
            self.base_url = "https://sandbox-api.reshadx.com/v1"

        # Create session with retry logic
        self.session = requests.Session()

        # Configure retries
        retry_strategy = Retry(
            total=max_retries,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "PUT", "DELETE", "OPTIONS", "TRACE", "POST"],
        )

        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

        # Set default headers
        self.session.headers.update(
            {
                "Content-Type": "application/json",
                "X-API-Key": self.api_key,
                "User-Agent": "ReshADX-Python-SDK/1.0.0",
            }
        )

    def set_access_token(self, token: str) -> None:
        """Set access token for authenticated requests"""
        self.access_token = token
        self.session.headers.update({"Authorization": f"Bearer {token}"})

    def clear_access_token(self) -> None:
        """Clear access token"""
        self.access_token = None
        if "Authorization" in self.session.headers:
            del self.session.headers["Authorization"]

    def request(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        json_data: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """Make HTTP request"""
        url = urljoin(self.base_url, path.lstrip("/"))

        try:
            response = self.session.request(
                method=method,
                url=url,
                params=params,
                json=json_data,
                timeout=self.timeout,
                **kwargs,
            )

            # Handle error responses
            if not response.ok:
                self._handle_error_response(response)

            # Parse JSON response
            data = response.json()

            # Return data field if present
            if isinstance(data, dict) and "data" in data:
                return data["data"]

            return data

        except requests.exceptions.Timeout:
            raise ReshADXError("Request timed out", "TIMEOUT_ERROR", 0)
        except requests.exceptions.ConnectionError:
            raise ReshADXError("Connection error", "NETWORK_ERROR", 0)
        except requests.exceptions.RequestException as e:
            raise ReshADXError(str(e), "NETWORK_ERROR", 0)

    def get(self, path: str, params: Optional[Dict[str, Any]] = None, **kwargs: Any) -> Any:
        """GET request"""
        return self.request("GET", path, params=params, **kwargs)

    def post(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """POST request"""
        return self.request("POST", path, json_data=json_data, **kwargs)

    def put(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """PUT request"""
        return self.request("PUT", path, json_data=json_data, **kwargs)

    def patch(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """PATCH request"""
        return self.request("PATCH", path, json_data=json_data, **kwargs)

    def delete(self, path: str, **kwargs: Any) -> Any:
        """DELETE request"""
        return self.request("DELETE", path, **kwargs)

    def _handle_error_response(self, response: requests.Response) -> None:
        """Handle error response"""
        try:
            error_data = response.json()
            if "error" in error_data:
                error = error_data["error"]
                raise ReshADXError(
                    message=error.get("message", "Unknown error"),
                    code=error.get("code", "UNKNOWN_ERROR"),
                    status_code=response.status_code,
                    details=error.get("details"),
                )
        except ValueError:
            pass

        # Fallback error
        raise ServerError(f"HTTP {response.status_code}: {response.text}")

File: sdks/typescript/src/utils/http.ts
/**
 * ReshADX TypeScript SDK - HTTP Client
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosError } from 'axios';
import axiosRetry from 'axios-retry';
import { ReshADXConfig, ApiError } from '../types';
import { ReshADXError } from './errors';

export class HttpClient {
  private client: AxiosInstance;
  private apiKey: string;
  private accessToken: string | null = null;

  constructor(config: ReshADXConfig) {
    this.apiKey = config.apiKey;

    const baseURL = config.baseUrl || this.getBaseUrl(config.environment || 'production');

    this.client = axios.create({
      baseURL,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.apiKey,
        'User-Agent': 'ReshADX-TypeScript-SDK/1.0.0',
      },
    });

    // Configure retry logic
    axiosRetry(this.client, {
      retries: config.maxRetries || 3,
      retryDelay: axiosRetry.exponentialDelay,
      retryCondition: (error) => {
        // Retry on network errors and 5xx server errors
        return (
          axiosRetry.isNetworkError(error) ||
          axiosRetry.isRetryableError(error) ||
          (error.response?.status !== undefined && error.response.status >= 500)
        );
      },
    });

    // Add request interceptor for authentication
    this.client.interceptors.request.use(
      (config) => {
        if (this.accessToken) {
          config.headers.Authorization = `Bearer ${this.accessToken}`;
        }
        return config;
      },
      (error) => Promise.reject(error)
    );

    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error: AxiosError<ApiError>) => {
        if (error.response?.data) {
          const apiError = error.response.data;
          throw new ReshADXError(
            apiError.error.message,
            apiError.error.code,
            error.response.status,
            apiError.error.details
          );
        }

        // Network or timeout error
        throw new ReshADXError(
          error.message || 'Network error occurred',
          'NETWORK_ERROR',
          0
        );
      }
    );
  }

  /**
   * Set access token for authenticated requests
   */
  setAccessToken(token: string): void {
    this.accessToken = token;
  }

  /**
   * Clear access token
   */
  clearAccessToken(): void {
    this.accessToken = null;
  }

  /**
   * GET request
   */
  async get<T>(url: string, config?: AxiosRequestConfig): Promise<T> {
    const response = await this.client.get<{ data: T }>(url, config);
    return response.data.data;
  }

  /**
   * POST request
   */
  async post<T>(url: string, data?: any, config?: AxiosRequestConfig): Promise<T> {
    const response = await this.client.post<{ data: T }>(url, data, config);
    return response.data.data;
  }

  /**
   * PUT request
   */
  async put<T>(url: string, data?: any, config?: AxiosRequestConfig): Promise<T> {
    const response = await this.client.put<{ data: T }>(url, data, config);
    return response.data.data;
  }

  /**
   * PATCH request
   */
  async patch<T>(url: string, data?: any, config?: AxiosRequestConfig): Promise<T> {
    const response = await this.client.patch<{ data: T }>(url, data, config);
    return response.data.data;
  }

  /**
   * DELETE request
   */
  async delete<T>(url: string, config?: AxiosRequestConfig): Promise<T> {
    const response = await this.client.delete<{ data: T }>(url, config);
    return response.data.data;
  }

  /**
   * Get base URL based on environment
   */
  private getBaseUrl(environment: 'sandbox' | 'production'): string {
    return environment === 'production'
      ? 'https://api.reshadx.com/v1'
      : 'https://sandbox-api.reshadx.com/v1';
  }
}
